---
title: "AssignASVs"
output: html_document
date: "2025-03-12"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      #send figures generated in this folder
                      fig.path = "../figures/02_AssignASVs/")
```


# Set up the environment
## Set seed
```{r set-seed}
set.seed(20020206)
```


## Load packages
```{r load-packages}
pacman::p_load(tidyverse, devtools, dada2,
               patchwork, DT, install = FALSE)

```


# Load filtered fastq files
```{r load-filtered-files}
# place filtered seq filed into a variable
filtered_fastqs_path <- "data/02_Filtered_fastqs/"

# Intuition check
filtered_fastqs_path

# Create forward vector
filtered_forward_reads <-
  list.files(filtered_fastqs_path, pattern = "_1_filtered.fastq.gz",
             full.names = TRUE)

# Check
filtered_forward_reads[1:5]

# Create forward vector
filtered_reverse_reads <-
  list.files(filtered_fastqs_path, pattern = "2_filtered.fastq.gz",
             full.names = TRUE)

# Check
filtered_reverse_reads[1:5]
```


# Sample Names
```{r sample-names}
# Create vector of sample names from the filenames 
sample_names <- sapply(strsplit(basename(filtered_forward_reads), "_"), `[`,1) 

# Intuition Check 
head(sample_names)

```


```{r learn-errors}
# Forward Reads
error_forward_reads <-
  learnErrors(filtered_forward_reads, multithread = 12)

# Plot
forward_error_plot <-
  plotErrors(error_forward_reads, nominalQ = TRUE) +
  labs(title = "Forward Read: Error Model")
forward_error_plot

# Reverse Reads
error_reverse_reads <-
  learnErrors(filtered_reverse_reads, multithread = 12)

# Plot
reverse_error_plot <-
  plotErrors(error_reverse_reads, nominalQ = TRUE) +
  labs(title = "Reverse Read: Error Model")
reverse_error_plot

# Look at the plots together
forward_error_plot + reverse_error_plot
```

## What are the current ASV lengths?
1. *What is the sequencing strategy?*
    - paired-end 2×150 bp Illumina HiSeq
2. *What's the total length of our starting amplicons?* 
    - 515F-806R region
    - 806−515=291bp
3. *What is the ASV length without primers?* 
    - 515F primer is 19 bp, and the 806R primer is 20 bp.
    - 291−19−20=252bp
4. *What is the length of the trimmed ASVs after `filterAndTrim()`?* 
    - trimLeft=0, no base at the beginning is removed, still 252bp
5. *What is the overlap between our forward and reverse reads?*
    - In the actual data, the forward sequence is 150bp and the reverse sequence is 230bp. After Trim, it is 140bp and 220bp. 
    - Overlap=220+140−252=108bp
    
<span style="color: red;">INTERPRETATION #1: The hypothesized ASV length for this dataset is 252 bp. This is based on the following considerations: (1) The sequencing strategy used was paired-end 2×150 bp Illumina HiSeq. (2) The total length of the starting amplicons, targeting the 515F-806R region, is 291 bp. (3) After removing the 515F (19 bp) and 806R (20 bp) primers, the expected ASV length without primers is 252 bp. (4) Since `trimLeft = 0`, no bases were removed from the start, and the ASV length remains 252 bp after `filterAndTrim()`. (5) The actual sequence lengths after trimming are 220 bp (forward) and 140 bp (reverse), resulting in an overlap of 108 bp, which is sufficient for accurate merging.</span>


# Infer ASVs
```{r infer-ASVs}
# Forward ASVs
dada_forward <-
  dada(filtered_forward_reads,
       err = error_forward_reads,
       multithread = 12)

# Take a look at the data
typeof(dada_forward)
dada_forward

# Reverse ASVs
dada_reverse <-
  dada(filtered_reverse_reads,
       err = error_reverse_reads,
       multithread = 20)

# Take a look at the data
typeof(dada_reverse)
dada_reverse[30]

```


# Merge Forward and Reverse ASVs
```{r merge-ASVs}
merged_ASVs <-
  mergePairs(dada_forward, filtered_forward_reads,
             dada_reverse, filtered_reverse_reads,
             verbose = TRUE)


# Evaluate the data output
typeof(merged_ASVs)
length(merged_ASVs)
names(merged_ASVs)

# Inspect further
head(merged_ASVs)
merged_ASVs$`022um-Control_R1_filtered.fastq.gz`


```


# Create a rough
```{r raw-ASV-count-table}
# Raw ASV
raw_ASV_table <- makeSequenceTable(merged_ASVs)

# Check
dim(raw_ASV_table)
typeof(raw_ASV_table)
class(raw_ASV_table)

# Write out raw_ASV_table
write.table(raw_ASV_table, file = "data/01_DADA2/raw_ASV_counts.tsv",
            sep = "\t", quote = FALSE, col.names = NA)


```


# Assess the ASV Length
```{r assess-ASV-length}
# Creating a table to inspect the distribution of ASV lengths
table(nchar(getSequences(raw_ASV_table)))

#Plot
plot_ASVLength_raw <-
  data.frame(ASV_Length = nchar(getSequences(raw_ASV_table))) %>%
  ggplot(aes(x = ASV_Length)) + 
  geom_histogram() +
  scale_x_continuous(limits = c(240, 265)) +
  labs(title = "Raw ASV Lengths",
       y = "Number of ASVs", x = "ASV Sequence Length (bps)")
print(plot_ASVLength_raw)
```
<span style="color: red;">INTERPRETATION #2: The expected length of the ASV is 252 bp, and in fact almost all ASVs are 253 bp, which is in line with expectations.


# Trim ASVs
```{r trim-ASVs}

# Only pull ASVs that have a length of 253 bps
raw_ASV_table_trimmed <-
  raw_ASV_table[,nchar(getSequences(raw_ASV_table)) == 253]
raw_ASV_table_trimmed

# Intuition Check
table(nchar(getSequences(raw_ASV_table_trimmed)))

#Plot
plot_ASVLength_trimmed <-
  data.frame(ASV_Length = nchar(getSequences(raw_ASV_table_trimmed))) %>%
  ggplot(aes(x = ASV_Length)) + 
  geom_histogram() +
  scale_x_continuous(limits = c(240, 265)) +
  labs(title = "Trimed Raw ASV Lengths",
       y = "Number of ASVs", x = "ASV Sequence Length (bps)")
print(plot_ASVLength_trimmed)
```
<span style="color: red;">INTERPRETATION #3: Remove all ASVs that are not 253 bp.


# Remove Chimeras
```{r remove-chimeras}
noChimeras_ASV_table <-
  removeBimeraDenovo(raw_ASV_table_trimmed,
                     method = "consensus",
                     multithread = 12,
                     verbose = TRUE)

# Structure of data
dim(noChimeras_ASV_table)
dim(raw_ASV_table_trimmed)

```

# What proportion is left of the sequences? 
# Chimera removal compared to trimming 
percRetained_chimerasTrimmed <- sum(noChimeras_ASV_table)/sum(raw_ASV_table_trimmed)
# Chimera removal compared to raw  
percRetained_chimerasRaw <-sum(noChimeras_ASV_table)/sum(raw_ASV_table)


```{r, plot_ASV_table}
# Plot it 
maxLength_ASV <- max(nchar(getSequences(noChimeras_ASV_table)), na.rm = TRUE)
plot_ASVLength_NoChimeras <- 
  data.frame(Seq_Length_NoChim = nchar(getSequences(noChimeras_ASV_table))) %>%
  ggplot(aes(x = Seq_Length_NoChim )) + 
  geom_histogram()+ 
  # include the x-axis scales
  scale_x_continuous(limits = c(240, maxLength_ASV + 12)) + 
  labs(title = "Post-Chimera ASV Lengths",
       y = "Number of ASVs \n (Post-Chimera Removal)", 
       x = "ASV Sequence Length (bps)")

# Show the plot
plot_ASVLength_NoChimeras 
```


```{r plot-ASVLengths, fig.height=5, fig.width=9}
print(plot_ASVLength_raw)
print(plot_ASVLength_trimmed)
print(plot_ASVLength_NoChimeras)
plot_ASVLength_raw + plot_ASVLength_trimmed + plot_ASVLength_NoChimeras + 
    plot_annotation(tag_levels = 'A')
```


```{r, ASV_check}
#trimmed/raw
prop_retained_trimming <- sum(raw_ASV_table_trimmed) / sum(raw_ASV_table)
print(paste("Proportion retained after trimming:", round(prop_retained_trimming, 4)))

#noChimeras/trimmed
prop_retained_chimera <- sum(noChimeras_ASV_table) / sum(raw_ASV_table_trimmed)
print(paste("Proportion retained after chimera removal:",
            round(prop_retained_chimera, 4)))

#noChimeras/raw
prop_retained_total <- sum(noChimeras_ASV_table) / sum(raw_ASV_table)
print(paste("Proportion retained after both trimming and chimera removal:",
            round(prop_retained_total, 4)))

```
<span style="color: red;">INTERPRETATION #4: Proportion retained after trimming: 0.9025; Proportion retained after chimera removal: 0.9649; Proportion retained after both trimming and chimera removal: 0.8709.Finally, 87.09% of the data can be used for subsequent analysis, which is a very healthy pipeline result.</span> 


```{r track-reads, fig.width=6, fig.height=4}
# A little function to identify number seqs 
getN <- function(x) sum(getUniques(x))

# Make the table to track the seqs 
track <- cbind(sapply(dada_forward, getN),
               sapply(dada_reverse, getN),
               sapply(merged_ASVs, getN),
               rowSums(noChimeras_ASV_table))

head(track)

# Update column names to be more informative (most are missing at the moment!)
colnames(track) <- c("denoisedF", "denoisedR", "merged", "nochim")
rownames(track) <- row.names(noChimeras_ASV_table)

# Generate a dataframe to track the reads through our DADA2 pipeline
track_counts_df <- 
  track %>%
  # make it a dataframe
  as.data.frame() %>%
  rownames_to_column(var = "sample_names")

# Now let's add a column for the number of ASVs
# First, intuition check that the samples match 
stopifnot(track_counts_df$sample_names == row.names(noChimeras_ASV_table))

# Now, let's add a new column with the number of ASVs
track_counts_df <- 
  track_counts_df %>%
  mutate(num_ASVs = rowSums(noChimeras_ASV_table > 1))

# Visualize it in table format 
DT::datatable(track_counts_df)

# Plot it!
track_counts_df %>%
  pivot_longer(denoisedF:nochim, names_to = "read_type", values_to = "num_reads") %>%
  mutate(read_type = fct_relevel(read_type, "denoisedF", "denoisedR", "merged", "nochim")) %>%
  ggplot(aes(x = read_type, y = num_reads, fill = read_type)) + 
  geom_line(aes(group = sample_names), color = "grey") + 
  geom_point(shape = 21, size = 3, alpha = 0.8) + 
  scale_fill_brewer(palette = "Spectral") + 
  labs(x = "Filtering Step", y = "Number of Sequences") + 
  theme_bw()
```


```{r numSeqsASV-plot, fig.height=2.5, fig.width=7}
plot_ReadDepth <- 
  track_counts_df %>%
  ggplot(aes(x = nochim)) + 
  geom_histogram() + 
  labs(x = "Total # of Sequences", y = "# of Samples") + 
  theme_bw()

# What is the ASV richness per sample? 
plot_ASVRichness <- 
  track_counts_df %>%
  ggplot(aes(x = num_ASVs)) + 
  geom_histogram() + 
  labs(x = "Total # of ASVs", y = "# of Samples") + 
  theme_bw()

# Now, let's look at the relationship of ASVs and Sequencing depth 
plot_ReadDepth_ASVRichness <- 
  track_counts_df %>%
  ggplot(aes(x = nochim, y = num_ASVs)) + 
  geom_point() + 
  labs(x = "Total # of Sequences", y = "# of ASVs") + 
  theme_bw()

# Show the plots together 
plot_ReadDepth + plot_ASVRichness + plot_ReadDepth_ASVRichness + 
    plot_annotation(tag_levels = 'A')
```

<span style="color: red;">INTERPRETATION #5: 1, The distribution appears roughly unimodal, centered around 10,000–25,000 reads per sample. 2, The ASV richness is also unimodal, mostly falling within 25 to 100 ASVs per sample.Some samples show higher ASV richness (~125 ASVs), while a few have low richness (< 25 ASVs). 3, There appears to be a positive correlation between sequencing depth and ASV richness, but the trend is not perfectly linear.</span>




